### DISCUSSION TITLE
Code not compiling

### DISCUSSION THREAD

**Discussion by denveronly** [txubelaxu/esphome-jk-bms#24] [Q&A]
Hey there, i had a code from the very beginning working excellently, but i want to rebuild my board and want to try out new code.
It doesnt compile for me. The code im using: it is latestest 1 master 7 slaves with very minor changes, components are downloaded from Txubelaxu git to local esphome folder

https://pastebin.com/HuA7pm7s

and errors received:

Failed config

sensor.jk_rs485_bms: [source /config/esphome/esptest.yaml:1078]
  platform: jk_rs485_bms
  jk_rs485_bms_id: bms2
  
  [cell_request_charge_voltage_time] is an invalid option for [sensor.jk_rs485_bms]. Did you mean [cell_average_voltage], [cell_delta_voltage]?
  cell_request_charge_voltage_time: 
    name: tbms2 cell request charge voltage time
  
  [cell_request_float_voltage_time] is an invalid option for [sensor.jk_rs485_bms]. Did you mean [cell_delta_voltage]?
  cell_request_float_voltage_time: 
    name: tbms2 cell request float voltage time
```

**Comment by denveronly** [comment]
ok i fixed it by commenting those [cell_request_charge_voltage_time] in sensor section, leaving the number section only.
Also i found that the example has a mistake:
word changes should be used!

```
switch:
  - platform: template
    name: "Broadcast changes to all BMSs"
    id: broadcast_active
    icon: "mdi:cast"
    restore_mode: RESTORE_DEFAULT_ON
    optimistic: true
    turn_on_action:
      - lambda: |-
          id(sniffer0).set_broadcast_changes_to_all_bms(true);  
    turn_off_action:
      - lambda: |-
          id(sniffer0).set_broadcast_changes_to_all_bms(false);   
```

**Reply by txubelaxu** [reply]
Yes, I moved some params from "sensor" to "number". This way, those items are now READ & WRITE.

@denveronly, what does "word changes should be used!"? The code is exaclty as you have written here. What is the mistake?

**Reply by denveronly** [reply]
in your code (0-7 Slaves)
![image](https://github.com/user-attachments/assets/1b30f9eb-8d92-4c80-8693-0e6dbd5905f6)
it doesnt compile because i receive and error:

![image](https://github.com/user-attachments/assets/0043a71a-83fb-406d-8e75-01e074f940ba)

so i updated the code as on top, added the word changes

**Reply by denveronly** [reply]
also, i found that the example has incorrect adresses for BMS.
Correct one should be, for bms0 - 0x00, for bms1 -0x01 etc.
![image](https://github.com/user-attachments/assets/224353fd-e3be-498f-981d-f25ba72d6d5b)

**Reply by txubelaxu** [reply]
I have just updated that example. Now it compiles OK.
The addresses start at 0x01 because all of them are slaves (7 slaves and 0 master). So, the ESP32 acts as master.
Remember that 0x00 is for master, so, that is why all jk addresses are > 0x00

If you do not want to ESP32 acting as master, then you can change the addresses to 0x00, 0x01, 0x02... But, in this case, not all the info is available for master bms, and none of the master bms params can be changed (read only). This is because of the RS485 limitation.

**Comment by denveronly** [comment]
can i also ask a question, how does the talk pin work?

**Reply by txubelaxu** [reply]
Well, there are different "UART to RS485" converters. RS485 is not full duplex, so, only one speaker can speak each time. So, converters need to know when they can speak. Some converters know that they can speak because of detecting changes at TX. Those do not need a "talk_pin" to work.

But another converters are not designed this way, so do not automatically detect that they can "speak" to the nerwork. They have a "read pin" and "write pin", that allow the speaker to signaling when "talk" and when "listen to".

If your converter automatically detects the TX, you do not need to use "talk_pin".

**Comment by affer** [comment]
denveronly: 

can you paste your working yaml? i alse get errors on compiling 0master_7slaves.yaml

pretty please :-D
